// radius/dist/m2dm/js/calculator/components/loan-breakdown.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define('radius/calculator/js/components/loan-breakdown', ['exports', 'react', 'prop-types'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('react'), require('prop-types'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.propTypes);
    global.radiusCalculatorJsComponentsLoanBreakdown = mod.exports;
  }
})(this, function (exports, _react, _propTypes) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _react2 = _interopRequireDefault(_react);

  var _propTypes2 = _interopRequireDefault(_propTypes);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  function LoanBreakdown(_ref) {
    var monthlyPayments = _ref.monthlyPayments,
        interestFee = _ref.interestFee,
        completionFee = _ref.completionFee,
        totalCost = _ref.totalCost;

    return _react2['default'].createElement(
      'div',
      null,
      _react2['default'].createElement(
        'h2',
        null,
        'Loan Breakdown'
      ),
      _react2['default'].createElement(
        'dl',
        null,
        _react2['default'].createElement(
          'dt',
          null,
          'Fixed monthly payments'
        ),
        _react2['default'].createElement(
          'dd',
          { 'data-test-monthly-payments': true },
          monthlyPayments
        ),
        _react2['default'].createElement(
          'dt',
          null,
          'Total interest'
        ),
        _react2['default'].createElement(
          'dd',
          { 'data-test-total-interest': true },
          interestFee
        ),
        _react2['default'].createElement(
          'dt',
          null,
          'Completion fee'
        ),
        _react2['default'].createElement(
          'dd',
          { 'data-test-completion-fee': true },
          completionFee
        ),
        _react2['default'].createElement(
          'dt',
          null,
          'Total cost of loan'
        ),
        _react2['default'].createElement(
          'dd',
          { 'data-test-total-cost': true },
          totalCost
        )
      )
    );
  }

  LoanBreakdown.propTypes = {
    monthlyPayments: _propTypes2['default'].string,
    interestFee: _propTypes2['default'].string,
    completionFee: _propTypes2['default'].string,
    totalCost: _propTypes2['default'].string
  };
  exports['default'] = LoanBreakdown;
});
