// radius/dist/m2dm/js/calculator/components/loan-amount-range-input.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define('radius/calculator/js/components/loan-amount-range-input', ['exports', 'react', 'prop-types'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('react'), require('prop-types'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.propTypes);
    global.radiusCalculatorJsComponentsLoanAmountRangeInput = mod.exports;
  }
})(this, function (exports, _react, _propTypes) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _react2 = _interopRequireDefault(_react);

  var _propTypes2 = _interopRequireDefault(_propTypes);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  var LoanAmountRangeInput = function LoanAmountRangeInput(_ref) {
    var maxLoanAmount = _ref.maxLoanAmount,
        minLoanAmount = _ref.minLoanAmount,
        loanAmount = _ref.loanAmount,
        updateLoanAmount = _ref.updateLoanAmount;
    return _react2['default'].createElement('input', {
      type: 'range',
      min: minLoanAmount,
      max: maxLoanAmount,
      value: loanAmount,
      onChange: function onChange(e) {
        return updateLoanAmount(parseInt(e.target.value, 10));
      }
    });
  };
  LoanAmountRangeInput.propTypes = {
    minLoanAmount: _propTypes2['default'].number,
    maxLoanAmount: _propTypes2['default'].number,
    loanAmount: _propTypes2['default'].number,
    updateLoanAmount: _propTypes2['default'].func
  };
  exports['default'] = LoanAmountRangeInput;
});
