// radius/dist/m2dm/js/calculator/components/terms.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define('radius/calculator/js/components/terms', ['exports', 'react', 'prop-types'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('react'), require('prop-types'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.propTypes);
    global.radiusCalculatorJsComponentsTerms = mod.exports;
  }
})(this, function (exports, _react, _propTypes) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _react2 = _interopRequireDefault(_react);

  var _propTypes2 = _interopRequireDefault(_propTypes);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  var Terms = function Terms(_ref) {
    var disabled = _ref.disabled,
        onTermSelected = _ref.onTermSelected,
        loanTerm = _ref.loanTerm,
        termLengths = _ref.termLengths;

    function createTerm(duration, isChecked) {
      return _react2['default'].createElement(
        'div',
        { key: duration },
        _react2['default'].createElement(
          'label',
          null,
          _react2['default'].createElement('input', {
            type: 'radio',
            name: 'term',
            value: duration,
            checked: isChecked,
            onChange: function onChange() {
              return onTermSelected(duration);
            }
          }),
          _react2['default'].createElement(
            'span',
            null,
            duration,
            ' months'
          )
        )
      );
    }

    return _react2['default'].createElement(
      'fieldset',
      { disabled: disabled, 'data-test-loan-terms': true },
      _react2['default'].createElement(
        'legend',
        null,
        'Loan term'
      ),
      termLengths.map(function (t) {
        return createTerm(t, t === loanTerm);
      })
    );
  };

  Terms.propTypes = {
    disabled: _propTypes2['default'].bool,
    loanTerm: _propTypes2['default'].number,
    onTermSelected: _propTypes2['default'].func,
    termLengths: _propTypes2['default'].array
  };

  exports['default'] = Terms;
});
