// radius/dist/m2dm/react/feature-list/index.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define('radius/feature-list/react/index', ['exports', 'react', 'classnames', 'prop-types', './FeatureListItem'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('react'), require('classnames'), require('prop-types'), require('./FeatureListItem'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.classnames, global.propTypes, global.FeatureListItem);
    global.radiusFeatureListReactIndex = mod.exports;
  }
})(this, function (exports, _react, _classnames, _propTypes, _FeatureListItem) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _react2 = _interopRequireDefault(_react);

  var _classnames2 = _interopRequireDefault(_classnames);

  var _propTypes2 = _interopRequireDefault(_propTypes);

  var _FeatureListItem2 = _interopRequireDefault(_FeatureListItem);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  function _defineProperty(obj, key, value) {
    if (key in obj) {
      Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
      });
    } else {
      obj[key] = value;
    }

    return obj;
  }

  function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
      throw new TypeError("Cannot call a class as a function");
    }
  }

  var _createClass = function () {
    function defineProperties(target, props) {
      for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
      }
    }

    return function (Constructor, protoProps, staticProps) {
      if (protoProps) defineProperties(Constructor.prototype, protoProps);
      if (staticProps) defineProperties(Constructor, staticProps);
      return Constructor;
    };
  }();

  function _possibleConstructorReturn(self, call) {
    if (!self) {
      throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }

    return call && (typeof call === "object" || typeof call === "function") ? call : self;
  }

  function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
      throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
    }

    subClass.prototype = Object.create(superClass && superClass.prototype, {
      constructor: {
        value: subClass,
        enumerable: false,
        writable: true,
        configurable: true
      }
    });
    if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
  }

  var FeatureList = function (_Component) {
    _inherits(FeatureList, _Component);

    function FeatureList() {
      _classCallCheck(this, FeatureList);

      return _possibleConstructorReturn(this, (FeatureList.__proto__ || Object.getPrototypeOf(FeatureList)).apply(this, arguments));
    }

    _createClass(FeatureList, [{
      key: 'render',
      value: function render() {
        var classes = (0, _classnames2['default'])('feature-list', _defineProperty({
          'u-color-white': this.props.contrast
        }, this.props.extraClasses, this.props.extraClasses));

        return _react2['default'].createElement(
          'ul',
          { className: classes },
          this.props.children
        );
      }
    }]);

    return FeatureList;
  }(_react.Component);

  FeatureList.Item = _FeatureListItem2['default'];
  exports['default'] = FeatureList;


  FeatureList.defaultProps = {
    contrast: false
  };

  FeatureList.propTypes = {
    children: _propTypes2['default'].oneOfType([_propTypes2['default'].arrayOf(_propTypes2['default'].element), _propTypes2['default'].element]).isRequired,
    contrast: _propTypes2['default'].bool,
    extraClasses: _propTypes2['default'].string
  };
});
