// radius/dist/m2dm/js/cookie-statement/index.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define('radius/cookie-statement/js/index', ['exports', '../cookies'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('../cookies'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.cookies);
    global.radiusCookieStatementJsIndex = mod.exports;
  }
})(this, function (exports, _cookies) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _cookies2 = _interopRequireDefault(_cookies);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  exports['default'] = function () {
    var selectorButtonDismiss = void 0;
    var cookieStatementPath = void 0;

    // Component configurable values
    var COOKIE_STATEMENT_NAME = 'FCCookieStatement';
    var COOKIE_STATEMENT_DAYS_FROM_NOW = 3653;

    // Values from the HTML markup
    var CLASS_ENABLED = 'is-enabled';
    var selectorCookieStatement = document.querySelector('[data-js-cookie-statement]');
    if (selectorCookieStatement) {
      selectorButtonDismiss = selectorCookieStatement.querySelector('[data-js-cookie-statement-cta]');
      cookieStatementPath = selectorCookieStatement.getAttribute('data-js-cookie-statement-path');

      if (!_cookies2['default'].getCookie(COOKIE_STATEMENT_NAME)) {
        selectorCookieStatement.className += ' ' + CLASS_ENABLED;
        selectorButtonDismiss.addEventListener('click', dismissCookieStatement);
      }
    }

    function dismissCookieStatement(event) {
      var cookieObj = {
        cookieName: COOKIE_STATEMENT_NAME,
        cookieValue: true,
        cookieDays: COOKIE_STATEMENT_DAYS_FROM_NOW,
        cookiePath: cookieStatementPath
      };
      event.preventDefault();
      selectorCookieStatement.className = selectorCookieStatement.className.replace(CLASS_ENABLED, '');
      _cookies2['default'].setCookie(cookieObj);
    }
  };
});
