// radius/dist/m2dm/js/header/header-with-animations.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define('radius/header/js/header-with-animations', ['exports', './header', './utils'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('./header'), require('./utils'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.header, global.utils);
    global.radiusHeaderJsHeaderWithAnimations = mod.exports;
  }
})(this, function (exports, _header, _utils) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _header2 = _interopRequireDefault(_header);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
      throw new TypeError("Cannot call a class as a function");
    }
  }

  var _createClass = function () {
    function defineProperties(target, props) {
      for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
      }
    }

    return function (Constructor, protoProps, staticProps) {
      if (protoProps) defineProperties(Constructor.prototype, protoProps);
      if (staticProps) defineProperties(Constructor, staticProps);
      return Constructor;
    };
  }();

  function _possibleConstructorReturn(self, call) {
    if (!self) {
      throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }

    return call && (typeof call === "object" || typeof call === "function") ? call : self;
  }

  var _get = function get(object, property, receiver) {
    if (object === null) object = Function.prototype;
    var desc = Object.getOwnPropertyDescriptor(object, property);

    if (desc === undefined) {
      var parent = Object.getPrototypeOf(object);

      if (parent === null) {
        return undefined;
      } else {
        return get(parent, property, receiver);
      }
    } else if ("value" in desc) {
      return desc.value;
    } else {
      var getter = desc.get;

      if (getter === undefined) {
        return undefined;
      }

      return getter.call(receiver);
    }
  };

  function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
      throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
    }

    subClass.prototype = Object.create(superClass && superClass.prototype, {
      constructor: {
        value: subClass,
        enumerable: false,
        writable: true,
        configurable: true
      }
    });
    if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
  }

  var HeaderWithAnimations = function (_Header) {
    _inherits(HeaderWithAnimations, _Header);

    function HeaderWithAnimations(headerEl) {
      var animationSpeed = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0.4;

      _classCallCheck(this, HeaderWithAnimations);

      var _this = _possibleConstructorReturn(this, (HeaderWithAnimations.__proto__ || Object.getPrototypeOf(HeaderWithAnimations)).call(this, headerEl));

      // Code compiled with babel6 in IE10 does not call the parent constructor
      // https://github.com/babel/babel/issues/3041
      _this.headerEl = headerEl;
      _this.isOpen = false;
      _this.overlayEl = _this.headerEl.querySelector(_header.OVERLAY_SELECTOR);
      _this.body = document.body;

      _this.inProgress = false;
      _this.config = {
        fromStyles: {
          opacity: 0
        },
        toStyles: {
          opacity: 1
        },
        animationSpeed: animationSpeed
      };
      return _this;
    }

    _createClass(HeaderWithAnimations, [{
      key: 'start',
      value: function start() {
        _get(HeaderWithAnimations.prototype.__proto__ || Object.getPrototypeOf(HeaderWithAnimations.prototype), 'start', this).call(this);

        if (this.hasOverlay) {
          this.overlayEl.style.transition = 'linear ' + this.config.animationSpeed + 's all';
        }
      }
    }, {
      key: 'showMenu',
      value: function showMenu() {
        var _this2 = this;

        if (this.inProgress || this.isOpen) {
          return;
        }

        this.isOpen = true;
        this.inProgress = true;

        this.freezeBody();
        this.overlayEl.style.display = 'block';

        Object.assign(this.overlayEl.style, this.config.fromStyles);

        (0, _utils.animate)(this.overlayEl, this.config.toStyles).then(function () {
          _this2.headerEl.classList.add(_header.IS_ACTIVE);
          _this2.inProgress = false;
        });
      }
    }, {
      key: 'hideMenu',
      value: function hideMenu() {
        var _this3 = this;

        if (this.inProgress || !this.isOpen) {
          return;
        }
        this.isOpen = false;
        this.inProgress = true;

        this.headerEl.classList.remove(_header.IS_ACTIVE);
        (0, _utils.animate)(this.overlayEl, this.config.fromStyles).then(function () {
          _this3.unFreezeBody();
          _this3.overlayEl.style.display = 'none';
          _this3.inProgress = false;
        });
      }
    }]);

    return HeaderWithAnimations;
  }(_header2['default']);

  exports['default'] = HeaderWithAnimations;
});
