// radius/dist/m2dm/js/calculator/index.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define('radius/calculator/js/index', ['exports', 'react', 'react-dom', 'react-redux', 'redux', './reducer', './components/calculator'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('react'), require('react-dom'), require('react-redux'), require('redux'), require('./reducer'), require('./components/calculator'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.reactDom, global.reactRedux, global.redux, global.reducer, global.calculator);
    global.radiusCalculatorJsIndex = mod.exports;
  }
})(this, function (exports, _react, _reactDom, _reactRedux, _redux, _reducer, _calculator) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _react2 = _interopRequireDefault(_react);

  var _reducer2 = _interopRequireDefault(_reducer);

  var _calculator2 = _interopRequireDefault(_calculator);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  /**
   *
   * section: M2DM/components/calculator
   * title: Calculator
   *
   * jsModule: calculator
   *
   * purpose: Use to allow the user to get an idea of how much a loan will cost them.
   *
   * markup: components/calculator/js/markup
   */

  var defaultState = {
    errorMessage: '',
    disabled: false,
    loanAmount: 250000,
    loanTerm: 36,
    minLoanAmount: 5000,
    maxLoanAmount: 1000000,
    locale: 'en-GB',
    termLengths: [6, 12, 24, 36, 48, 60]
  };

  exports['default'] = function () {
    if (document.querySelector('[data-js-calculator]')) {
      var store = (0, _redux.createStore)(_reducer2['default'], defaultState);

      (0, _reactDom.render)(_react2['default'].createElement(
        _reactRedux.Provider,
        { store: store },
        _react2['default'].createElement(_calculator2['default'], null)
      ), document.querySelector('[data-js-calculator]'));
    }
  };
});
