// radius/dist/m2dm/js/calculator/containers/loan-amount-numeric-input.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define('radius/calculator/js/containers/loan-amount-numeric-input', ['exports', 'react-redux', '../actions', '../components/loan-amount-numeric-input'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('react-redux'), require('../actions'), require('../components/loan-amount-numeric-input'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.reactRedux, global.actions, global.loanAmountNumericInput);
    global.radiusCalculatorJsContainersLoanAmountNumericInput = mod.exports;
  }
})(this, function (exports, _reactRedux, _actions, _loanAmountNumericInput) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _loanAmountNumericInput2 = _interopRequireDefault(_loanAmountNumericInput);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  function mapStateToProps(_ref) {
    var loanAmount = _ref.loanAmount,
        locale = _ref.locale,
        minLoanAmount = _ref.minLoanAmount,
        maxLoanAmount = _ref.maxLoanAmount,
        errorMessage = _ref.errorMessage;

    return {
      loanAmount: loanAmount,
      locale: locale,
      minLoanAmount: minLoanAmount,
      maxLoanAmount: maxLoanAmount,
      errorMessage: errorMessage
    };
  }

  function mapDispatchToProps(dispatch) {
    return {
      updateLoanAmount: function updateLoanAmount(value) {
        dispatch((0, _actions.updateLoanAmountAction)(value));
      }
    };
  }

  exports['default'] = (0, _reactRedux.connect)(mapStateToProps, mapDispatchToProps)(_loanAmountNumericInput2['default']);
});
