// radius/dist/m2dm/js/calculator/containers/loan-amount-range-input.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define('radius/calculator/js/containers/loan-amount-range-input', ['exports', 'react-redux', '../actions', '../components/loan-amount-range-input'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('react-redux'), require('../actions'), require('../components/loan-amount-range-input'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.reactRedux, global.actions, global.loanAmountRangeInput);
    global.radiusCalculatorJsContainersLoanAmountRangeInput = mod.exports;
  }
})(this, function (exports, _reactRedux, _actions, _loanAmountRangeInput) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _loanAmountRangeInput2 = _interopRequireDefault(_loanAmountRangeInput);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  function mapStateToProps(_ref) {
    var loanAmount = _ref.loanAmount,
        minLoanAmount = _ref.minLoanAmount,
        maxLoanAmount = _ref.maxLoanAmount;

    return {
      loanAmount: loanAmount,
      minLoanAmount: minLoanAmount,
      maxLoanAmount: maxLoanAmount
    };
  }

  function mapDispatchToProps(dispatch) {
    return {
      updateLoanAmount: function updateLoanAmount(value) {
        dispatch((0, _actions.updateLoanAmountAction)(value));
      }
    };
  }

  exports['default'] = (0, _reactRedux.connect)(mapStateToProps, mapDispatchToProps)(_loanAmountRangeInput2['default']);
});
