// radius/dist/m2dm/js/calculator/containers/terms.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define('radius/calculator/js/containers/terms', ['exports', 'react-redux', '../actions', '../components/terms'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('react-redux'), require('../actions'), require('../components/terms'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.reactRedux, global.actions, global.terms);
    global.radiusCalculatorJsContainersTerms = mod.exports;
  }
})(this, function (exports, _reactRedux, _actions, _terms) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _terms2 = _interopRequireDefault(_terms);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  function mapStateToProps(_ref) {
    var disabled = _ref.disabled,
        errorMessage = _ref.errorMessage,
        loanTerm = _ref.loanTerm,
        termLengths = _ref.termLengths;

    return {
      disabled: disabled,
      errorMessage: errorMessage,
      loanTerm: loanTerm,
      termLengths: termLengths
    };
  }

  function mapDispatchToProps(dispatch) {
    return {
      onTermSelected: function onTermSelected(term) {
        return dispatch((0, _actions.updateTermAction)(term));
      }
    };
  }

  exports['default'] = (0, _reactRedux.connect)(mapStateToProps, mapDispatchToProps)(_terms2['default']);
});
