// radius/src/current/js/sidenav.js
import { debounce } from 'lodash';
import utils from './utils';

export default (options = {}) => {
  const sidenavEl = document.querySelector('.js-r-sidenav');
  if (! sidenavEl) {
    return;
  }

  const toggleEl = sidenavEl.querySelector('.js-r-sidenav-toggle');
  const BREAKPOINT = 640;
  const ACTIVE_CLASS = 'is-active';
  const STICKY_CLASS = 'is-sticky';
  const RATIO = 1.62;
  const sidenavInitialOffset = sidenavEl.getBoundingClientRect().top + window.pageYOffset;
  const config = options;

  if (config.scrollSpy === void 0) { config.scrollSpy = false; }

  if (sidenavEl.length !== 0 && toggleEl.length !== 0) {
    bindLinks();
    bindToggle();
    bindScroll();
    bindResize();
  }

  function bindLinks() {
    utils.addEventListener(sidenavEl, 'click', changeLink);
  }

  function changeLink(e) {
    if (e.target.tagName.toLowerCase() !== 'a') {
      return;
    }

    toggleFull();
    setActiveLink(e.target);
  }

  function setActiveLink(el) {
    if (!config.scrollSpy) {
      return;
    }

    const activeEl = sidenavEl.getElementsByClassName(ACTIVE_CLASS)[0];

    utils.removeClass(activeEl, ACTIVE_CLASS);

    utils.addClass(el, ACTIVE_CLASS);
  }

  function bindToggle() {
    utils.addEventListener(toggleEl, 'click', (e) => {
      e.stopPropagation();
      toggleFull();
    });
  }

  function bindScroll() {
    utils.addEventListener(window, 'scroll', debounce(() => {
      if (window.innerWidth > BREAKPOINT) {
        const windowScroll = window.pageYOffset;
        const fromTop = window.innerHeight / (RATIO + 1);

        if (sidenavInitialOffset < (windowScroll + fromTop)) {
          if (!utils.hasClass(sidenavEl, STICKY_CLASS)) {
            utils.addClass(sidenavEl, STICKY_CLASS);
          }

          sidenavEl.style.top = `${fromTop}px`;
        } else {
          utils.removeClass(sidenavEl, STICKY_CLASS);
          sidenavEl.style.top = '';
        }
      }
    }, 33));
  }

  function bindResize() {
    window.onresize = debounce(() => {
      if (window.innerWidth < BREAKPOINT) {
        sidenavEl.style.top = '';

        utils.removeClass(sidenavEl, STICKY_CLASS);
      }
    }, 100);
  }

  function toggleFull() {
    if (utils.hasClass(sidenavEl, ACTIVE_CLASS)) {
      utils.removeClass(sidenavEl, ACTIVE_CLASS);
    } else {
      utils.addClass(sidenavEl, ACTIVE_CLASS);
    }
  }
};
