// radius/src/m2dm/components/calculator/js/containers/loan-breakdown.js
import { connect } from 'react-redux';
import LoanBreakdown from '../components/loan-breakdown';
import calculateAmortisedValue from '../calculate-amortisation';

function formatCurrency(value, locale) {
  let currency;
  switch (locale) {
    case 'en-GB':
      currency = 'GBP';
      break;
    case 'en-US':
      currency = 'USD';
      break;
    default:
      currency = 'EUR';
  }

  return value.toLocaleString(locale, { style: 'currency', currency, minimumFractionDigits: 0 });
}

function mapStateToProps({
  loanTerm, loanAmount, locale, errorMessage,
}) {
  if (errorMessage) {
    const errorString = '-';
    return {
      monthlyPayments: errorString,
      interestFee: errorString,
      completionFee: errorString,
      totalCost: errorString,
    };
  }
  const {
    monthlyPayments, interestFee, completionFee, totalCost,
  } = calculateAmortisedValue(loanAmount, loanTerm);

  return {
    monthlyPayments: formatCurrency(monthlyPayments, locale),
    interestFee: formatCurrency(interestFee, locale),
    completionFee: formatCurrency(completionFee, locale),
    totalCost: formatCurrency(totalCost, locale),
  };
}

export default connect(mapStateToProps)(LoanBreakdown);
